aptos-move/block-executor/src/executor.rs

```diff
@@ -685,40 +744,57 @@ where
                     &mut scheduler_task,
                     last_input_output,
                     shared_commit_state,
+                    base_view,
+                    shared_counter,
+                    &executor,
+                    block,
                 );
                 scheduler.queueing_commits_mark_done();
             }
 
             drain_commit_queue();
 
             scheduler_task = match scheduler_task {
-                SchedulerTask::ValidationTask(txn_idx, incarnation, wave) => Self::validate(
-                    txn_idx,
-                    incarnation,
-                    wave,
-                    last_input_output,
-                    versioned_cache,
-                    scheduler,
-                ),
+                SchedulerTask::ValidationTask(txn_idx, incarnation, wave) => {
+                    let valid = Self::validate(txn_idx, last_input_output, versioned_cache);
+                    Self::update_on_validation(
+                        txn_idx,
+                        incarnation,
+                        valid,
+                        wave,
+                        last_input_output,
+                        versioned_cache,
+                        scheduler,
+                    )
+                },
                 SchedulerTask::ExecutionTask(
                     txn_idx,
                     incarnation,
                     ExecutionTaskType::Execution,
                 ) => {
                     let mut shared_commit_state_guard = shared_commit_state.acquire();
                     let (_, _, maybe_error) = shared_commit_state_guard.dereference_mut();
-                    Self::execute(
+                    let InternalExecuteResult {
+                        record_succeeded,
+                        updates_outside,
+                    } = Self::execute(
                         txn_idx,
                         incarnation,
                         block,
                         last_input_output,
                         versioned_cache,
-                        scheduler,
                         &executor,
                         base_view,
                         ParallelState::new(versioned_cache, scheduler, shared_counter),
-                        maybe_error,
-                    )
+                    );
+                    if record_succeeded {
```

Comment:
hm, seeing how you refactored validation, maybe something similar can prettify the execution flow as well, i.e. the record handling (both success and failure) seems to be shareable between here and the commit re-execution? (so a method like update_on_validation.. update_on_execution)